### ISSUE TITLE
Pixelpipe: High CPU load even opencl is working

### ISSUE THREAD

**Issue by AxelG-DE** [darktable-org/darktable#2404]
**Describe the bug**
I have had this experience in the past as well in official branch. It disappered for me, when I added a second GPU (2xGTX1060 G1 6GB).

Now it came back to happen, that regardless of opencl seems to work very fine, there remains a lot of CPU-load (core i9 9900). That getting even worse, when the second preview window is in use.

I know, I AmaZe is slow, however, there was a time, it worked without loading the CPU that much...

"25,496388 [pixelpipe_process] [full] using device 0
25,499503 [dev_pixelpipe] took 0,003 secs (0,031 CPU) initing base buffer [full]
25,502229 [opencl_profiling] profiling device 1 ('GeForce GTX 1060 6GB'):
25,502241 [opencl_profiling] spent  0,0041 seconds in [Write Image (from host to device)]
25,502245 [opencl_profiling] spent  0,0001 seconds in rawprepare_1f
25,502248 [opencl_profiling] spent  0,0001 seconds in whitebalance_1f
25,502268 [opencl_profiling] spent  0,0001 seconds in highlights_1f_lch_bayer
25,502270 [opencl_profiling] spent  0,0103 seconds in [Read Image (from device to host)]
25,502272 [opencl_profiling] spent  0,0003 seconds in colorin_unbound
25,502275 [opencl_profiling] spent  0,0007 seconds in [Copy Image (on device)]
25,502276 [opencl_profiling] spent  0,0004 seconds in colorspaces_transform_lab_to_rgb_matrix
25,502278 [opencl_profiling] spent  0,0005 seconds in exposure
25,502280 [opencl_profiling] spent  0,0004 seconds in blendop_mask_rgb
25,502281 [opencl_profiling] spent  0,0004 seconds in blendop_rgb
25,502284 [opencl_profiling] spent  0,0003 seconds in basecurve_lut
25,502286 [opencl_profiling] spent  0,0004 seconds in colorspaces_transform_rgb_matrix_to_lab
25,502288 [opencl_profiling] spent  0,0004 seconds in tonecurve
25,502290 [opencl_profiling] spent  0,0003 seconds in sharpen_hblur
25,502292 [opencl_profiling] spent  0,0002 seconds in sharpen_vblur
25,502294 [opencl_profiling] spent  0,0003 seconds in sharpen_mix
25,502296 [opencl_profiling] spent  0,0006 seconds in colorout
25,502298 [opencl_profiling] spent  0,0199 seconds totally in command queue (with 0 events missing)
25,502320 [dev_process_preview] pixel pipeline processing took 0,072 secs (0,474 CPU)
25,505873 [pixelpipe_process] [preview] using device 1
25,505899 [dev_process_preview] pixel pipeline processing took 0,000 secs (0,000 CPU)
25,506089 [dev_pixelpipe] took 0,007 secs (0,011 CPU) processed `Raw-Schwarz-/Weißpunkt' on GPU, blended on GPU [full]
25,507887 [dev_pixelpipe] took 0,002 secs (0,002 CPU) processed `Weißabgleich' on GPU, blended on GPU [full]
25,510824 [dev_pixelpipe] took 0,003 secs (0,001 CPU) processed `Spitzlicht-Rekonstruktion' on GPU, blended on GPU [full]
25,651407 [pixelpipe_process] [preview] using device 1
25,651456 [dev_pixelpipe] took 0,000 secs (0,000 CPU) initing base buffer [preview]
25,653283 [dev_pixelpipe] took 0,002 secs (0,027 CPU) processed `Raw-Schwarz-/Weißpunkt' on GPU, blended on GPU [preview]
25,653968 [dev_pixelpipe] took 0,001 secs (0,014 CPU) processed `Weißabgleich' on GPU, blended on GPU [preview]
25,654492 [dev_pixelpipe] took 0,001 secs (0,001 CPU) processed `Spitzlicht-Rekonstruktion' on GPU, blended on GPU [preview]
25,684815 [dev_pixelpipe] took 0,030 secs (0,430 CPU) processed `Entrastern' on CPU, blended on CPU [preview]
25,689418 [dev_pixelpipe] took 0,005 secs (0,019 CPU) processed `Eingabefarbprofil' on GPU, blended on GPU [preview]
image colorspace transform Lab-->RGB took 0,001 secs (0,003 GPU) [exposure ]
25,695225 [dev_pixelpipe] took 0,184 secs (2,369 CPU) processed `Entrastern' on CPU, blended on CPU [full]
25,700867 [dev_pixelpipe] took 0,011 secs (0,025 CPU) processed `Belichtung' on GPU, blended on GPU [preview]
25,701696 [dev_pixelpipe] took 0,001 secs (0,002 CPU) processed `Belichtung 1' on GPU, blended on GPU [preview]"

**To Reproduce**
1. Go to 'darkroom'
2. Click on 'switch on several modules like demosaic-AmaZe, exposure with L-mask, second over all exposure, Highlight reconstruction LCZ'
3. Keep moving white balance slider and observe CPU-load, e.g. with gkrellm

**Expected behavior**
Almost no CPU load, as it was previously.

**Screenshots**

**Platform (please complete the following information):**
 - OS: [gentoo linux] with nvidia-drivers-418.56 und gentoo-kernel linux-5.0.7-gentoo
 - Version [2.7.0+998~g959466989-dirty]

**Comment by AxelG-DE** [comment]
now attached I have a log file, where I have an empty pixelpipe (just std. with basecurve) and moved the white-balance continuously. It caused me 20% cpu-load of my core i9

I hope I am not wrong, but I seem to remember, it was way less cpu-load previously

[dt-log.20190414b.txt](https://github.com/darktable-org/darktable/files/3077756/dt-log.20190414b.txt)

quick scanning this file, I didn't see any weird unlike above, where I think this line proves my observations: " [dev_pixelpipe] took 0,184 secs (2,369 CPU) processed "

dt is at 2.7.0+1006~g64bd45ad2-dirty

**Comment by AxelG-DE** [comment]
Just now went back to dt-2.6.2 (had to go with a fresh ~/.config/darktable since I edited productively with 2.7 but tried also my existing darktablerc)
- in **dt-2.6.2** moving WB slider cause CPU-load of **~5%** (more intuitive, since no lag)
- back in 2.7.0+1007~g2c02989e8-dirty moving WB slider cause CPU-load of **~20%** 
(much less intuitive, since noticable lag)

**Comment by aurelienpierre** [comment]
% of CPU load are not a valid metric. Actually, you want your CPU to be used at 100%. Please run `darktable -d perf` and compare running times.

**Comment by AxelG-DE** [comment]
@aurelienpierre

I see, what you're saying, e.g. exporting 4s on 50% load is worth than 2s on 100% CPU load.

Fully agree to that. 

However, what I tried to explain above is slightly different.... 
I experience within dt 2.7 a higher CPU load when I move WB slider and hence a way less fluent, more sticky user experience. When I go back to 2.6.2 with the same picture and same history stack, the response to WB slider feels WAY more responsive and hence less hindering me in my anyway improbable creativity.

Actually it feels somehow metric to mee. 5% CPU load *feels* 4 times fluent than 20% CPU load.

BTW: Difficult to use -d perf for comparison as you get too much output when you keep moving sliders for e.g. 3-4 seconds. 

I'm afraid @darix will track me yet  again for not concisely describing my thoughts :blush: :smile:

**Comment by AxelG-DE** [comment]
I created those two log files with the following setup:
-copied a folder and renamed it
-deleted all .xmp on that testfolder
-moved ~/.config/darktable
-opened dt with the command: "/opt/darktable/bin/darktable -d opencl -d perf 2&>dt-27_20190416.log" (and similar for 2.6.2
-imported that new test folder
-opened one image (no history stack, base curve and sharpening enabled)
-moved slider 6 times fully right and fully left

I clearly felt the less responsive means less performant on 2.7 compared to 2.6.2 And guess what, also observing "watch nvidia-smi" and also for opencl the GPU load seems lower on 2.6.2 even though the response is more fluent...

I did quickly look into the files with something like this (also with "Entrastern" which is demosaic and "Weißabgleich" which is whitebalance):
"colordiff dt-262_20190416.log dt-27_20190416.log | grep Raw-Schwarz-/Weißp" and I **must admit,** what I see there **seems not supporting** my *complain*. Even more I ask you, to look into it open minded. 

I do have a quite powerful machine here but feel very noticeable performance loss in dt 2.7, which is special pity, because this topic [pr2365](https://github.com/darktable-org/darktable/pull/2365) somehow costs double performance (very logic by this approach)

"2.7.0+1022~g6a9574be3-dirty"
[dt-27_20190416.log](https://github.com/darktable-org/darktable/files/3086363/dt-27_20190416.log)

"2.6.2"
[dt-262_20190416.log](https://github.com/darktable-org/darktable/files/3086364/dt-262_20190416.log)

**Comment by AxelG-DE** [comment]
BTW: 
I do have activated *opencl_synch_cache=active module* but does not really help here

[pr2367](https://github.com/darktable-org/darktable/pull/2367)

**Comment by aurelienpierre** [comment]
What I meant was timing the whole pipe or using external CPU loads is meaningless to us because we need to find out what specific modules have caused this regression, and how bad it is.

There is a new feature that allows modules to work in RGB space in the middle of the Lab part of the pixelpipe, but this comes at the expense of some additional conversions. That could be an explanation. Anyway, we need more specific info.

Also, there is a scheduler that chooses if the modules should use the GPU or fallback to the CPU if the GPU latency is too high, so, depending on what happens in there, that could be an explanation too.

TL;DR please send the complete painstaking log.

**Comment by AxelG-DE** [comment]
@aurelienpierre got ya, thanks

I find "my" phenomenon :smile: also on an almost empty history stack. 

TL;DR ? I'm a bit lost
Swapped the keys and finally you meant lighttable and darkroom? 

Complete painstaking you mean "-d all"
I'll do so later

**Comment by aurelienpierre** [comment]
"-d perf" will do ;-)

**Comment by AxelG-DE** [comment]
@aurelienpierre 
above in https://github.com/darktable-org/darktable/issues/2404#issuecomment-483795424 I do have this:
"2.7.0+1022~g6a9574be3-dirty"
dt-27_20190416.log

it was done with "-d opencl -d perf"

Something not good with that log-file? Something I could do better?

**Comment by aurelienpierre** [comment]
Looking at individual modules timings, it's on par. "Entrasten" (demoisaicing I guess ?) is even a bit faster on 2.7. Your whole pixelpipe goes from 76 ms (2.6.2) to 82 ms (2.7). Am I missing something ?

**Comment by AxelG-DE** [comment]
That's what I tried to explain above too, when I said: 

> I did quickly look into the files with something like this (also with "Entrastern" which is demosaic and "Weißabgleich" which is whitebalance):
"colordiff dt-262_20190416.log dt-27_20190416.log | grep Raw-Schwarz-/Weißp" and I must admit, what I see there seems not supporting my complain. Even more I ask you, to look into it open minded.

Those logs I did with an almost empty pixelpipe to avoid we get distracted by other modules. Actually and logically the effect gets stronger with more resource hungry modules involved, but it is noticeable also on an empty pixel pipe.

How did I find it: 
Indeed one better change WB at the beginning of a processing. However, after upgrading my machine to core i9 I enjoyed like hell, that even at the end, I could re-adjust WB and it was still very fluent in reaction. No lag at all. And now with dt 2.7 I am somehow bag to the status of my Core i7 with 6x 2.8GHz....

**Comment by AxelG-DE** [comment]
Maybe I found something...
one Photo from play raw (Thomas' Tulips) I right-clicked on WB, which was at 2000 at that moment and I entered 25000. and what I can see from that action, it runs the pixelpipe two times and you can also see, that one action lasts ~1.5s

- I leave the line 11.46729s in, so you can see, there was nothing else before 40s....
- I donno "insert code" is a good idea, I may edit

`11,467297 [dev_process_image] pixel pipeline processing took 0,000 secs (0,000 CPU)
40,959922 [dev] took 0,000 secs (0,000 CPU) to load the image.
40,985559 [dev_pixelpipe] took 0,026 secs (0,007 CPU) processed `Weißabgleich' on GPU, blended on GPU [preview]
40,985984 [dev_pixelpipe] took 0,000 secs (0,001 CPU) processed `Spitzlicht-Rekonstruktion' on GPU, blended on GPU [preview]
40,987648 [dev_pixelpipe] took 0,002 secs (0,003 CPU) processed `Entrastern' on GPU, blended on GPU [preview]
40,989084 [dev_pixelpipe] took 0,001 secs (0,002 CPU) processed `Eingabefarbprofil' on GPU, blended on GPU [preview]
image colorspace transform Lab-->RGB took 0,001 secs (0,001 GPU) [basecurve ]
40,992751 [dev_pixelpipe] took 0,004 secs (0,005 CPU) processed `Basiskurve' on GPU, blended on GPU [preview]
image colorspace transform RGB-->Lab took 0,001 secs (0,001 GPU) [sharpen ]
40,997329 [dev_pixelpipe] took 0,005 secs (0,005 CPU) processed `schärfen' on GPU, blended on GPU [preview]
40,999104 [dev_pixelpipe] took 0,002 secs (0,002 CPU) processed `Ausgabefarbprofil' on GPU, blended on GPU [preview]
41,005313 [dev_pixelpipe] took 0,006 secs (0,030 CPU) processed `Gamma' on CPU, blended on CPU [preview]
41,006592 [dev_pixelpipe] took 0,047 secs (0,063 CPU) processed `Weißabgleich' on GPU, blended on GPU [full]
41,009669 [dev_pixelpipe] took 0,003 secs (0,027 CPU) processed `Spitzlicht-Rekonstruktion' on GPU, blended on GPU [full]
image colorspace transform RGB-->RGB took 0,006 secs (0,050 CPU) [final histogram]
41,025370 [dev_process_preview] pixel pipeline processing took 0,065 secs (0,135 CPU)
41,025454 [dev_process_preview] pixel pipeline processing took 0,000 secs (0,000 CPU)
41,052120 [dev_pixelpipe] took 0,042 secs (0,084 CPU) processed `Entrastern' on GPU, blended on GPU [full]
41,059889 [dev_pixelpipe] took 0,008 secs (0,008 CPU) processed `Eingabefarbprofil' on GPU, blended on GPU [full]
image colorspace transform Lab-->RGB took 0,002 secs (0,002 GPU) [basecurve ]
41,065185 [dev_pixelpipe] took 0,005 secs (0,007 CPU) processed `Basiskurve' on GPU, blended on GPU [full]
image colorspace transform RGB-->Lab took 0,002 secs (0,002 GPU) [sharpen ]
41,072609 [dev_pixelpipe] took 0,007 secs (0,011 CPU) processed `schärfen' on GPU, blended on GPU [full]
41,075297 [dev_pixelpipe] took 0,003 secs (0,003 CPU) processed `Ausgabefarbprofil' on GPU, blended on GPU [full]
41,084703 [dev_pixelpipe] took 0,009 secs (0,051 CPU) processed `Gamma' on CPU, blended on CPU [full]
41,084724 [dev_process_image] pixel pipeline processing took 0,125 secs (0,258 CPU)
41,515401 [dev] took 0,000 secs (0,000 CPU) to load the image.
41,515467 [dev_process_image] pixel pipeline processing took 0,000 secs (0,000 CPU)
  `

**Comment by aurelienpierre** [comment]
So maybe an UI issue then ? Something with GTK or Cairo lagging ? What is your GTK lib version ?

**Comment by AxelG-DE** [comment]
gtk+ is 3.24.4-r1 and 2.24.32-31
cairo is 1.16.0-r3

**Comment by AxelG-DE** [comment]
Dear @aurelienpierre , 

sorry to bother you again with this topic. 
- I think you are right. At the end of the day it is something with the UI
- I tried in XFCE and experience the same
- accidently I found it in tone-curve as well and the -perf output is very suspicious to me everything took allegedly 0.000s but at the end, just switch it on (not that severe tone-curve) takes ~0.5s (see below)
- when I go back to 2.6.2 that effect is gone 
- dt-2.7 with new ~/.config/darktable also shows this effect

here is my -d perf output from just switching a tonecurve on (or off)
68,128125 [dev_process_preview] pixel pipeline processing took 0,000 secs (0,000 CPU)
68,227898 [dev] took 0,000 secs (0,000 CPU) to load the image.
68,227959 [dev_process_image] pixel pipeline processing took 0,000 secs (0,000 CPU)
68,677023 [dev] took 0,000 secs (0,000 CPU) to load the image.
68,677078 [dev_process_image] pixel pipeline processing took 0,000 secs (0,000 CPU)

Thank you very much for staying with me on this.
If I can do anything to analyse more, pls. let me know

**Comment by AxelG-DE** [comment]
Dear all,

@aurelienpierre said at the beginning, CPU-load is not a proper measure to describe performance issues. By that time I thought, a short push from an expert and things will run smooth again. Now I believe, I found something bigger. Let's hope I am not wrong...

I think, I made it to nail down performance issues of darkroom, utilizing a spreadsheet. Also I assume, those are also closely linked to this issue [#2064](https://github.com/darktable-org/darktable/issues/2064)

Gimme some time to complete the spreadsheet and make it more presentable, then:
- I'd open a new issue
- subsequently I will close this one on my own

Cheers
 Axel
